### PR TITLE
fix: prevent focus from moving beyond toDate and fromDate

### LINKED ISSUE #1320: Focus navigation breaks with disabled days
### Bug description

When navigating the calendar with the keyboard the focus breaks when the day to focus is disabled

### To reproduce 

[Code sandbox: https://codesandbox.io/s/react-day-picker-8-vl4ry](https://codesandbox.io/s/react-day-picker-8-forked-yk2ez?file=/src/App.tsx)

#### Steps

Focus the calendar and move with the arrows until a disabled day is met. The focus get lost.

### Expected behavior

The focus should get to the next (or previous?) focusable (not disabled) day. If there are no days to navigate to, the focus should stay in the last navigable day.

### Additional notes

Test should check all the keyboard combinations. Consider the specific use cases (e.g. navigate end of week when end of week is disabled).

### PR DESCRIPTION
**Pull request by kimamula**
Prevent focus from moving beyond toDate and fromDate to fix #1320

### CODE REVIEW

**Comment by kimamula** [comment]
What about always enabling the behavior I implemented behind `disallowFocusOnDisabledDates`?

**Comment by gpbl** [comment]
> What about always enabling the behavior I implemented behind `disallowFocusOnDisabledDates`?

That behavior is native in the browser already when we rollback #1449. I'd say is a better solution!

**Comment by kimamula** [comment]
The behavior I implemented here prevents the issue described in #1320 from happening.

**Comment by kimamula** [comment]
@gpbl reverted #1451 and removed `disallowFocusOnDisabledDates` and `disallowFocusOnDisabledNavigation` props

**Comment by kimamula** [comment]
@gpbl FYI, @react-aria/datepicker behaves similarly to what I implemented here when `minValue` and/or `maxValue` is set.

https://react-spectrum.adobe.com/react-aria/useDatePicker.html#minimum-and-maximum-values

**Comment by kimamula** [comment]
@gpbl is it possible to estimate when you will be able to take a look at this? Thanks!

**Comment by gpbl** [comment]
> The behavior I implemented here prevents the issue described in #1320 from happening.

Thanks @kimamula, I will check it out. 

Still we need to fix #1320 in a PR ad-hoc, after reverting #1451. The fix should focus only on the fix and have updated tests.

If you want to help:

- revert #1451 in a different PR
- open a PR with the changes that would fix #1320. 
 - this should include one or more tests that confirm the issue is fixed

If you have urgency with this issue, consider to downgrade to `v8.0.4` (if the [recent fixes](https://github.com/gpbl/react-day-picker/releases) aren't bugging you). Also [sponsoring the project](https://github.com/sponsors/gpbl) would help to push forward its development. Thanks!

**Comment by kimamula** [comment]
@gpbl thanks! Reverted #1451 in #1476 and updated this PR to focus on fixing #1320.

**Comment by kimamula** [comment]
@gpbl could you kindly review this PR? Downgrading to v8.0.4 doesn't work for us since #1320 is also not acceptable. Thanks!

**Comment by gpbl** [comment]
I could finally try out this PR, but there's an issue: when meeting a disabled day, the user needs to press the arrow button twice to skip the disabled day: 



What was the issue and how does this PR fix it? Some context would help for a quicker review!

**Comment by gpbl** [comment]
I think we'd better close #1483 for a easier development on our local machine. Also, it seems the tests don't cover the issue. The test should confirm that the disabled days are skipped when moving with the keyboard navigation. 

Thanks for your help!

**Comment by kimamula** [comment]
Thank you for your review!!

I believe this PR fixes the issue shown in the CodeSandbox example you provided in #1320. I didn't know about `disabled` prop and didn't address the issue you pointed out (I updated the title of this PR to avoid confusion).

It seems difficult and will take a lot of time to determine how to move the focus when encountering a `disabled`. It would be great if the issue could be made an item to be addressed in a separate PR. What do you think?

@react-aria/datepicker renders disabled dates without actually "disabling" them. This way, it does not have this issue as well as the issue I described in #1449.

**Comment by kimamula** [comment]
I found that the issue described in your [comment](https://github.com/gpbl/react-day-picker/pull/1468#issuecomment-1186016263) is not specific to `disabled` but also happens with `hidden`, for which the solution adopted in @react-aria/datepicker doesn't work.

I still hope this issue could be addressed in a separate PR, but here are my thoughts on how to solve the issue:

- `focusStartOfWeek`: move to the first focusable (not disabled or hidden) day
- `focusStartOfWeek`: move to the last focusable day
- `focusDayBefore`, `focusWeekBefore`, `focusMonthBefore`, `focusYearBefore`:
  1. Move to the previous day, week, month, or year if it is selectable.
  2. If it is not, move to the nearest day to 1.
  3. If 2 is the currently focused day, move to the previous focusable day.
  4. If 3 does not exist, stay at the currently focused day.
- `focusDayAfter`, `focusWeekAfter`, `focusMonthAfter`, `focusYearAfter`:
  1. Move to the next day, week, month, or year if it is selectable.
  2. If it is not, move to the nearest day to 1.
  3. If 2 is the currently focused day, move to the next focusable day.
  4. If 3 does not exist, stay at the currently focused day.

```ts
// pseudo code
const moveFocus = (addFn: typeof addDays, after: boolean) => {
    if (!focusedDay) return;
    let newFocusedDay = addFn(focusedDay, after ? 1 : -1);
    if (!isFocusable(newFocusedDay)) {
        newFocusedDay = findNearestFocusableDayTo(newFocusedDay);
        if (isSameDay(focusedDay, newFocusedDay) {
            newFocusedDay = after ? findNextFocusableDay(focusedDay) : findPreviousFocusableDay(focusedDay);
        }
    }
    if (!newFocusedDay) return;
    navigation.goToDate(newFocusedDay, focusedDay);
    focus(newFocusedDay);
};

const focusDayBefore = () => moveFocus(addDays, false);
const focusDayAfter = () => moveFocus(addDays, true);
const focusWeekBefore = () => moveFocus(addWeeks, false);
const focusWeekAfter = () => moveFocus(addWeeks, true);
const focusMonthBefore = () => moveFocus(addMonths, false);
const focusMonthAfter = () => moveFocus(addMonths, true);
const focusYearBefore = () => moveFocus(addYears, false);
const focusYearAfter = () => moveFocus(addYears, true);

const focusStartOfWeek = () => {
    const newFocusedDay = getFocusalbeDaysInWeek(focusedDay)?.[0];
    if (!newFocusedDay) return;
    navigation.goToDate(newFocusedDay, focusedDay);
    focus(newFocusedDay);
};
const focusEndOfWeek = () => {
    const newFocusedDay = getFocusalbeDaysInWeek(focusedDay)?.pop();
    if (!newFocusedDay) return;
    navigation.goToDate(newFocusedDay, focusedDay);
    focus(newFocusedDay);
};
```

**Comment by McSam27** [comment]
Any ETA on this branch being merged in?

**Comment by kimamula** [comment]
@gpbl is there anything I can do to get this PR merged?

**Comment by gpbl** [comment]
> @gpbl is there anything I can do to get this PR merged?

Thanks, there are few bugs surfacing listed here https://github.com/gpbl/react-day-picker/issues, it would be nice some help there too! 

Now that https://github.com/gpbl/react-day-picker/issues/1483 is closed it should be easier to debug this PR. 

@McSam27 please don't ask for ETA thanks I'm working on this issue in my free-time 🙏

File: packages/react-day-picker/src/contexts/Focus/FocusContext.tsx [hunk 1]
```diff
@@ -83,75 +86,32 @@ export function FocusProvider(props: { children: ReactNode }): JSX.Element {
     setFocusedDay(date);
   };
 
-  const focusDayBefore = () => {
+  const moveFocus = (addFn: typeof addDays, after: boolean) => {
```
Comment (gpbl) [inline packages/react-day-picker/src/contexts/Focus/FocusContext.tsx:89]: I like this change - I'd better move it in an external function tho... I will update the code when merged.

**Review by gpbl — APPROVED** [review]
Looks great @kimamula ! Thanks for your patience – I want to update in a second PR some code and ask for a review 🙏

**Pull request merged**